### PR TITLE
Implement Encoder and EncoderBLock

### PR DESCRIPTION
**Pull request by jackfreem**
## Description

Implements the EncoderBlock and then sets up the structure to duplicate this N times (which I am calling the Encoder)

## Checklist

Please ensure the following before submitting the PR:

- [ ] PR is Squashed
- [x] I have added/updated relevant documentation (e.g., README.md, inline docstrings).
- [x] I have written tests for my changes.
- [x] All tests pass locally.

## How to Test

` pytest .\tests\`
Note: We're getting lots of warnings on the softmax. I left a comment about this, but will plan to fix in a fast follow

## Notes for Reviewers
- Am going to put up Decoder & DecoderBlock in a branch off of this PR shortly.

### CODE REVIEW

File: src/layers/scaled_dot.py [hunk 1]
```diff
@@ -38,6 +38,7 @@ def forward(
             # -9e15 to approx neg infinity -> 0s out in softmax -> (exp(-9e15) ≈ 0)
             x = x.masked_fill(mask == 0, -9e15)
         
+        # Apply softmax along the last dimension (-1) to normalize the scores
         attn = f.softmax(x)
```
Comment (jackfreem) [inline src/layers/scaled_dot.py:42]: I think this is what we need to do, but it gives me some errors I'll need to work through. Going to try to fix with a fast follow, but leave the comment as reminder.

File: tests/blocks/test_encoder_block.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import torch
+import unittest
+from src.blocks.encoder_block import EncoderBlock
+
+class TestEncoderBlock(unittest.TestCase):
```
Comment (bradford-gill) [inline tests/blocks/test_encoder_block.py:5]: please add case to test that `mask = None` works

**Review by bradford-gill — APPROVED** [review]
i only found one main thing i though should be covered, please add the simple test. 

Approving

File: tests/blocks/test_encoder_block.py (hunk 2 above)
Comment (jackfreem) [reply tests/blocks/test_encoder_block.py:5]: Going to do this in a follow-up PR to prevent merge conflicts.
